### ISSUE TITLE
Docker + Kubernetes + Hazelcast = Cache Freezes ? 

### ISSUE THREAD

**Issue by theobolo** [hazelcast/hazelcast#13310]
Hello, i'm working on a projet using Hazelcast and Hazelcast Jet for our company. We are developping micro-services applications in JAVA within a Kubernetes environment, and we are using Hazelcast to manage application clusters and cache across our replicated instances.

Since few weeks we discovered a weird problem about the cache reponse time, that we are only facing when Hazelcast is deployed into our Kubernetes clusters. Frequently hazelcast is taking more than 2 seconds to get 1000 users from cache, it sounds like cache is freezing or something like that, but we have no clues about it ...

To isolate the problem we developped a small app, client/server using the lastest version of Hazelcast (3.10) available here : https://github.com/theobolo/hazelcast-test

This application is creating 10 000 users using a Java DAO class and load them into memory (server side), other hazelcast instances (clients) are getting thoses users and then are trying to get randomly 1000 users accounts from the Hazelcast Cache and from 2 other methods to compare the results : 

 - Memory Cache
 - TCP Cache
 - Hazelcast Cache

On a local environment, by just running the applications JAR file, we have pretty good performances with 1 client and 1 server : 

![image](https://user-images.githubusercontent.com/7281773/41529206-9e413a7e-72ec-11e8-85e3-fe47319d72e8.png)

Even with 3 clients and 3 servers , we don't have any trouble or freezy thing. 

But when we are running our test application on a Kubernetes Cluster, using a custom Docker image, we have this strange problem : 

![image](https://user-images.githubusercontent.com/7281773/41527310-01cb8654-72e7-11e8-8d1d-5189ebaee868.png)

Periodically we are experiencing Hazelcast cache taking more than 2 seconds to retrieve the 1000 users and we don't know why. 

We are using a very simple Dockerfile : https://github.com/theobolo/hazelcast-test/blob/master/Dockerfile
Using a OpenJDK Java Alpine based image, with no stuff in it except the Jar file. 
We are using a Kubernetes cluster (version 1.10.0) with 3 masters and 3 workers (8 VCpus / 36Go Ram) on Azure, all the workers are in the same DataCenter and the same RessourceGroup.

The problem occurs basically with 1 client and 1 server, and is more visible when we add more servers and more clients, for exemple 3 clients and 3 servers : 

![image](https://user-images.githubusercontent.com/7281773/41527586-ecc6ccd6-72e7-11e8-8f1e-214da0499fcf.png)

For the moment the test application is using Zookeeper to discover the other instances.

We are totally lost about that issue because we tried that app on different Kubernetes environnements : Minikube / Kubernetes on Azure / GCE / AWS and we always face the same problem. Even with different hardware behind our clusters (basically over sized clusters)
We also tried to use the previous Hazelcast version 3.9 witch ended in the same way. 

You can also find the `run.sh` script used to launch our Jar into the Docker container right here : https://github.com/theobolo/hazelcast-test/blob/master/run.sh

Even by desactivating all the run.sh stuff like JAVA_OPTS and JMX endpoint we are facing the same reponse time problem. 

Do you have any ideas about that ?

**Comment by vincentwuhrlin** [comment]
Hello,

just to add some information (i am the CTO, i am working with Theo), you will find a screenshot of our web application in production (not the test application sources, but the problem is identical) which is a Hazelcast Client embedded and connected to an other application which is Hazecast Member embedded and use cache.

For the moment we have no near cache.

So, you can notice on screenshot that the time to get from cache is very random : sometimes 3ms, other time 3s, so 1000x bigger ! We have a little cache (< 100 elements in cache and size < 1ko).

**We don't have this problem locally, only in a Kubernetes environment** (even in minikube in a single machine)

We have no idea for the moment, and we have disabled today the use of Hazelcast cache and replaced by database call (completely illogical !) because our customers have disastrous response time on our dashborad web application when cache is used.

Have you some experiences of Hazelcast application in a Kubernetes environment ?

Thank you very much.

Sincerely,
Vincent

![capture d ecran 2018-06-20 a 18 47 31](https://user-images.githubusercontent.com/16716356/41672937-8ea876c4-74bb-11e8-98e5-d3d645acd835.png)

**Comment by enesakar** [comment]
Are clients in the same kubernetes cluster?
Have you tested with same docker images (Alpine) but without kubernetes?

**Comment by pveentjer** [comment]
Can you enable diagnostics:

http://docs.hazelcast.org/docs/latest-development/manual/html/Management/Diagnostics/Enabling_Diagnostics_Logging.html

It would also be helpful to have some profiling information so we can see what happens when there is such a stall.

On both client and server and include in this ticket? Or otherwise send them to peter at hazelcast dot com

**Comment by theobolo** [comment]
Hello @enesakar, i didn't try yet just using Docker without Kubernetes, i'm on it !

@pveentjer I enabled Hazelcast Diagnostics on both client and server side, for the moment it's not showing up any troubles ... 

I started the client and the server without any JMX / JAVA_OPTS options, just with the Diagnostics properties : 

Here you can see 2 Diag files from the CLIENT : 
https://gist.github.com/theobolo/bdc65b13ceae339b89219ca02bb837a6

And here you can see 2 Diag files from SERVER (Member) : 
https://gist.github.com/theobolo/4977fa6d009524bd04e8addbb1d25798

Client kubernetes Logs : https://gist.github.com/theobolo/24a0e82ec1d335af2dd30b485a277809
Server kubernetes Logs : https://gist.github.com/theobolo/24ec7a08f732b4f0e39cd6049cd0bb10

On the client Side you can see some ~1.7seconds request time.

**Comment by cforce** [comment]
I believe bridging/routing or even bonding is routinly used in virtualized environments and you should disable LRO to be on the safe side. See this VMWare knowledgebase article.

Also the Intel ixgbe driver documentation has a good read:

WARNING:  The ixgbe driver compiles by default with the LRO (Large Receive
Offload) feature enabled.  This option offers the lowest CPU utilization for
receives, but is completely incompatible with *routing/ip forwarding* and
*bridging*.  If enabling ip forwarding or bridging is a requirement, it is
necessary to disable LRO using compile time options as noted in the LRO
section later in this document.  The result of not disabling LRO when combined
with ip forwarding or bridging can be low throughput or even a kernel panic.
and from the same document:

Do Not Use LRO When Routing or Bridging Packets
-----------------------------------------------
Due to a known general compatibility issue with LRO and routing, do not use
LRO when routing or bridging packets.
Bottom Line:

Disable LRO in virtualized environment or when you use network interface bonding. Enable it only when you absolutely have to and only after extensive testing

LRO is a per-interface setting. If LRO is turned on for an interface, the CPU will offload the processing of incoming packets (received on that interface) to the NIC. To try to reproduce or get rid this bug, LRO must be turned on for the public/external interface on the Docker host.

https://github.com/hazelcast/hazelcast-docker/issues/63

the docker host network interface which runs pods need to set "ethtool -K eth0 lro off"
